File: app/src/main/java/org/stepic/droid/ui/fragments/SettingsFragment.java
package org.stepic.droid.ui.fragments;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;

import org.stepic.droid.R;
import org.stepic.droid.base.FragmentBase;
import org.stepic.droid.notifications.model.NotificationType;
import org.stepic.droid.ui.custom.BetterSwitch;
import org.stepic.droid.ui.dialogs.AllowMobileDataDialogFragment;
import org.stepic.droid.ui.dialogs.VideoQualityDialog;

import butterknife.BindString;
import butterknife.BindView;

public class SettingsFragment extends FragmentBase implements AllowMobileDataDialogFragment.Callback {

    public static SettingsFragment newInstance() {
        return new SettingsFragment();
    }

    @BindView(R.id.fragment_settings_wifi_enable_switch)
    BetterSwitch wifiLoadSwitch;

    @BindView(R.id.fragment_settings_external_player_switch)
    BetterSwitch externalPlayerSwitch;

    @BindView(R.id.video_quality_view)
    View videoQuality;

    @BindView(R.id.fragment_settings_notification_learn_switch)
    BetterSwitch notificationLearnSwitch;

    @BindView(R.id.fragment_settings_notification_comment_switch)
    BetterSwitch notificationCommentSwitch;

    @BindView(R.id.fragment_settings_notification_review_switch)
    BetterSwitch notificationReviewSwitch;

    @BindView(R.id.fragment_settings_notification_teaching_switch)
    BetterSwitch notificationTeachingSwitch;

    @BindView(R.id.fragment_settings_notification_other_switch)
    BetterSwitch notificationOtherSwitch;


    @BindView(R.id.fragment_settings_notification_vibration_switch)
    BetterSwitch notificationVibration;

    @BindView(R.id.fragment_settings_notification_sound_switch)
    BetterSwitch notificationSound;

    @BindView(R.id.fragment_settings_calendar_widget_switch)
    BetterSwitch calendarWidgetSwitch;

    @BindView(R.id.fragment_settings_keep_screen_on_switch)
    BetterSwitch keepScreenOnSwitch;

    @BindView(R.id.storage_management_button)
    View storageManagementButton;

    @BindView(R.id.fragment_settings_discounting_policy_switch)
    BetterSwitch discountingPolicySwitch;

    @BindString(R.string.version)
    String versionPrefix;

    @BindString(R.string.clear_cache_title)
    String clearCacheTitle;

    @BindString(R.string.kb)
    String kb;

    @BindString(R.string.mb)
    String mb;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_settings, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setUpNotificationVibration();

        setUpNotifications();

        setUpSound();

        wifiLoadSwitch.setChecked(!sharedPreferenceHelper.isMobileInternetAlsoAllowed());//if first time it is true

        externalPlayerSwitch.setChecked(userPreferences.isOpenInExternal());

        externalPlayerSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                userPreferences.setOpenInExternal(isChecked);
            }
        });

        calendarWidgetSwitch.setChecked(userPreferences.isNeedToShowCalendarWidget());

        calendarWidgetSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean isChecked) {
                userPreferences.setNeedToShowCalendarWidget(isChecked);
            }
        });

        keepScreenOnSwitch.setChecked(userPreferences.isKeepScreenOnSteps());
        keepScreenOnSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                userPreferences.setKeepScreenOnSteps(isChecked);
            }
        });

        discountingPolicySwitch.setChecked(userPreferences.isShowDiscountingPolicyWarning());

        discountingPolicySwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                userPreferences.setShowDiscountingPolicyWarning(isChecked);
            }
        });


        wifiLoadSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean newCheckedState) {
                if (wifiLoadSwitch.isUserTriggered()) {
                    if (newCheckedState) {
                        //wifi only
                        onMobileDataStateChanged(false);
                    } else {
                        //wifi and mobile internet
                        wifiLoadSwitch.setChecked(true);
                        AllowMobileDataDialogFragment dialogFragment = AllowMobileDataDialogFragment.Companion.newInstance();
                        dialogFragment.setTargetFragment(SettingsFragment.this, 0);
                        if (!dialogFragment.isAdded()) {
                            dialogFragment.show(getFragmentManager(), null);
                        }
                    }

                }
            }
        });


        final DialogFragment videoDialog = new VideoQualityDialog();
        videoQuality.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (!videoDialog.isAdded()) {
                    videoDialog.show(getFragmentManager(), null);
                }
            }
        });

        storageManagementButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                screenManager.showStorageManagement(getActivity());
            }
        });

    }

    private void setUpNotificationVibration() {
        notificationVibration.setChecked(userPreferences.isVibrateNotificationEnabled());
        notificationVibration.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                userPreferences.setVibrateNotificationEnabled(isChecked);
            }
        });
    }

    private void setUpSound() {
        notificationSound.setChecked(userPreferences.isSoundNotificationEnabled());
        notificationSound.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                userPreferences.setNotificationSoundEnabled(isChecked);
            }
        });
    }

    @Override
    public void onDestroyView() {
        keepScreenOnSwitch.setOnCheckedChangeListener(null);
        discountingPolicySwitch.setOnCheckedChangeListener(null);
        calendarWidgetSwitch.setOnCheckedChangeListener(null);
        wifiLoadSwitch.setOnCheckedChangeListener(null);
        externalPlayerSwitch.setOnCheckedChangeListener(null);
        notificationLearnSwitch.setOnCheckedChangeListener(null);
        notificationCommentSwitch.setOnCheckedChangeListener(null);
        notificationTeachingSwitch.setOnCheckedChangeListener(null);
        notificationOtherSwitch.setOnCheckedChangeListener(null);
        notificationReviewSwitch.setOnCheckedChangeListener(null);
        notificationVibration.setOnCheckedChangeListener(null);
        notificationSound.setOnCheckedChangeListener(null);
        storageManagementButton.setOnClickListener(null);
        super.onDestroyView();
    }

    private void storeMobileState(boolean isMobileAllowed) {
        sharedPreferenceHelper.setMobileInternetAndWifiAllowed(isMobileAllowed);
    }

    private void setUpNotifications() {
        notificationLearnSwitch.setChecked(userPreferences.isNotificationEnabled(NotificationType.learn));
        notificationLearnSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                userPreferences.setNotificationEnabled(NotificationType.learn, isChecked);
            }
        });

        notificationCommentSwitch.setChecked(userPreferences.isNotificationEnabled(NotificationType.comments));
        notificationCommentSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                userPreferences.setNotificationEnabled(NotificationType.comments, isChecked);
            }
        });

        notificationReviewSwitch.setChecked(userPreferences.isNotificationEnabled(NotificationType.review));
        notificationReviewSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                userPreferences.setNotificationEnabled(NotificationType.review, isChecked);
            }
        });

        notificationTeachingSwitch.setChecked(userPreferences.isNotificationEnabled(NotificationType.teach));
        notificationTeachingSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                userPreferences.setNotificationEnabled(NotificationType.teach, isChecked);
            }
        });

        notificationOtherSwitch.setChecked(userPreferences.isNotificationEnabled(NotificationType.other));
        notificationOtherSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                userPreferences.setNotificationEnabled(NotificationType.other, isChecked);
            }
        });

    }

    @Override
    public void onMobileDataStateChanged(boolean isMobileAllowed) {
        wifiLoadSwitch.setChecked(!isMobileAllowed);
        storeMobileState(isMobileAllowed);
    }
}

File: app/src/main/java/org/stepic/droid/ui/fragments/SettingsFragment.kt
package org.stepic.droid.ui.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.fragment_settings.*
import org.stepic.droid.R
import org.stepic.droid.base.FragmentBase
import org.stepic.droid.ui.dialogs.AllowMobileDataDialogFragment
import org.stepic.droid.ui.dialogs.CoursesLangDialog
import org.stepic.droid.ui.dialogs.VideoQualityDialog
import org.stepik.android.view.font_size_settings.ui.dialog.ChooseFontSizeDialogFragment
import ru.nobird.android.view.base.ui.extension.showIfNotExists

class SettingsFragment : FragmentBase(), AllowMobileDataDialogFragment.Callback {
    companion object {
        fun newInstance(): SettingsFragment = SettingsFragment()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.fragment_settings, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        nullifyActivityBackground()

        notificationActionButton.setOnClickListener {
            screenManager.showNotificationSettings(activity)
        }

        fragmentSettingsWifiEnableSwitch.isChecked = !sharedPreferenceHelper.isMobileInternetAlsoAllowed//if first time it is true

        fragmentSettingsExternalPlayerSwitch.isChecked = userPreferences.isOpenInExternal

        fragmentSettingsExternalPlayerSwitch.setOnCheckedChangeListener { _, isChecked -> userPreferences.isOpenInExternal = isChecked }

        fragmentSettingsCalendarWidgetSwitch.isChecked = userPreferences.isNeedToShowCalendarWidget

        fragmentSettingsCalendarWidgetSwitch.setOnCheckedChangeListener { _, isChecked -> userPreferences.isNeedToShowCalendarWidget = isChecked }

        fragmentSettingsKeepScreenOnSwitch.isChecked = userPreferences.isKeepScreenOnSteps
        fragmentSettingsKeepScreenOnSwitch.setOnCheckedChangeListener { _, isChecked -> userPreferences.isKeepScreenOnSteps = isChecked }

        fragmentSettingsAdaptiveMode.isChecked = userPreferences.isAdaptiveModeEnabled
        fragmentSettingsAdaptiveMode.setOnCheckedChangeListener { _, isChecked -> userPreferences.isAdaptiveModeEnabled = isChecked }

        fragmentSettingsDiscountingPolicySwitch.isChecked = userPreferences.isShowDiscountingPolicyWarning

        fragmentSettingsDiscountingPolicySwitch.setOnCheckedChangeListener { _, isChecked -> userPreferences.isShowDiscountingPolicyWarning = isChecked }

        fragmentSettingsAutoplay.isChecked = userPreferences.isAutoplayEnabled
        fragmentSettingsAutoplay.setOnCheckedChangeListener { _, isChecked -> userPreferences.isAutoplayEnabled = isChecked }

        fragmentSettingsWifiEnableSwitch.setOnCheckedChangeListener { _, newCheckedState ->
            if (fragmentSettingsWifiEnableSwitch.isUserTriggered) {
                if (newCheckedState) {
                    //wifi only
                    onMobileDataStateChanged(false)
                } else {
                    //wifi and mobile internet
                    fragmentSettingsWifiEnableSwitch.isChecked = true
                    val dialogFragment = AllowMobileDataDialogFragment.newInstance()
                    dialogFragment.setTargetFragment(this@SettingsFragment, 0)
                    if (!dialogFragment.isAdded) {
                        dialogFragment.show(requireFragmentManager(), null)
                    }
                }
            }
        }


        videoQualityView.setOnClickListener {
            val videoDialog = VideoQualityDialog.newInstance(forPlaying = false)
            if (!videoDialog.isAdded) {
                videoDialog.show(requireFragmentManager(), null)
            }
        }

        videoPlayingQualityView.setOnClickListener {
            val videoDialog = VideoQualityDialog.newInstance(forPlaying = true)
            if (!videoDialog.isAdded) {
                videoDialog.show(requireFragmentManager(), null)
            }
        }

        storageManagementButton.setOnClickListener { screenManager.showStorageManagement(activity) }

        langWidgetActionButton.setOnClickListener {
            CoursesLangDialog.newInstance().show(requireFragmentManager(), null)
        }

        fontSizeSettingsButton.setOnClickListener {
            ChooseFontSizeDialogFragment
                .newInstance()
                .showIfNotExists(requireFragmentManager(), ChooseFontSizeDialogFragment.TAG)
        }
    }

    override fun onDestroyView() {
        fragmentSettingsKeepScreenOnSwitch.setOnCheckedChangeListener(null)
        fragmentSettingsDiscountingPolicySwitch.setOnCheckedChangeListener(null)
        fragmentSettingsCalendarWidgetSwitch.setOnCheckedChangeListener(null)
        fragmentSettingsWifiEnableSwitch.setOnCheckedChangeListener(null)
        fragmentSettingsExternalPlayerSwitch.setOnCheckedChangeListener(null)
        storageManagementButton.setOnClickListener(null)
        notificationActionButton.setOnClickListener(null)
        super.onDestroyView()
    }

    private fun storeMobileState(isMobileAllowed: Boolean) {
        sharedPreferenceHelper.setMobileInternetAndWifiAllowed(isMobileAllowed)
    }

    override fun onMobileDataStateChanged(isMobileAllowed: Boolean) {
        fragmentSettingsWifiEnableSwitch.isChecked = !isMobileAllowed
        storeMobileState(isMobileAllowed)
    }
}